xrdp daemon should not warn about running as root or try to do rootful things when systemd controls uid/gid

### xrdp version

0.10.4, 0.10.80

### Detailed xrdp version, build options

~~~
xrdp 0.10.80
  A Remote Desktop Protocol Server.
  Copyright (C) 2004-2025 Jay Sorg, Neutrino Labs, and all contributors.
  See https://github.com/neutrinolabs/xrdp for more information.

  Configure options:
      --build=x86_64-redhat-linux
      --host=x86_64-redhat-linux
      --program-prefix=
      --disable-dependency-tracking
      --prefix=/usr
      --exec-prefix=/usr
      --bindir=/usr/bin
      --sbindir=/usr/sbin
      --sysconfdir=/etc
      --datadir=/usr/share
      --includedir=/usr/include
      --libdir=/usr/lib64
      --libexecdir=/usr/libexec
      --localstatedir=/var
      --runstatedir=/run
      --sharedstatedir=/var/lib
      --mandir=/usr/share/man
      --infodir=/usr/share/info
      --enable-fuse
      --enable-pixman
      --enable-painter
      --enable-vsock
      --enable-ipv6
      --enable-openh264
      --enable-utmp
      --with-socketdir=/run/xrdp
      --with-imlib2
      build_alias=x86_64-redhat-linux
      host_alias=x86_64-redhat-linux
      CC=gcc
      CFLAGS=-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Wno-complain-wrong-lang -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64-v3 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2
      LDFLAGS=-Wl,-z,relro -Wl,--as-needed  -Wl,-z,pack-relative-relocs -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1 -specs=/usr/lib/rpm/redhat/redhat-package-notes
      CXX=g++
      CXXFLAGS=-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64-v3 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2
      LT_SYS_LIBRARY_PATH=/usr/lib64:
      PKG_CONFIG_PATH=:/usr/lib64/pkgconfig:/usr/share/pkgconfig
~~~


### Operating system & version

Rocky Linux 10.0 (Red Quartz), Rocky Linux 9.6 (Blue Onyx), Rocky Linux 8.10 (Green Obsidian), Fedora Linux 42 (KDE Plasma Desktop Edition)

### Installation method

Doesn't matter but DNF, some from package repository, some self-built via rpmbuild.

### Which backend do you use?

xorgxrdp but it should not matter

### What desktop environment do you use?

Xfce but it should not matter

### Environment xrdp running on

Both physical and VM, not using GPU:s (but Nvidia T4 (Rocky 10 vm and physical Rocky 8 server), L40s (Rocky 9 server), GeForce MX550 (Fedora laptop))

### What's your client?

N/A, ssh

### Area(s) with issue?

Other

### Steps to reproduce

1. Run on any systemd-based distro.
2. Install and activate xrdp
3. do `systemctl edit xrdp` and add:

```
[Service]
User=xrdp
Group=xrdp
LogsDirectory=xrdp
```

4. edit `/etc/xrdp/xrdp.ini` and set
```
LogFile=xrdp/xrdp.log
```

(optional, changes behaviour) 5. Edit `/etc/xrdp/xrdp.ini` and set

```
runtime_user=
runtime_group=
```

### ✔️ Expected Behavior

xrdp daemon starts as normal, running as the `xrdp` user, without complaining or incorrectly claiming to be running as root.

### ❌ Actual Behavior

1. If step 5 was skipped, xrdp crashes with error:
```
Aug 28 09:29:05 wayrocky10 xrdp[162956]: [ERROR] Unable to init groups for 'xrdp' [Operation not permitted]
```
because it tries to do initgroups without privileges.

2. If in step 5 you set the user/group to nothing, then a spurious warning is printed in the journal:

```
"You are running xrdp as root. This is not safe."
```

### Anything else?

The issue seems to be in xrdp/xrdp.c, specifically the function [check_drop_privileges](https://github.com/neutrinolabs/xrdp/blob/devel/xrdp/xrdp.c#L423):

```
static int
check_drop_privileges(struct xrdp_startup_params *startup_params)
{
    int rv = 1;
    const char *user = startup_params->runtime_user;
    const char *group = startup_params->runtime_group;

    if (user[0] == '\0' && group[0] == '\0')
    {
        // Allow this for now
        LOG(LOG_LEVEL_ALWAYS,
            "You are running xrdp as root. This is not safe.");
        rv = 0;
    }
    else if (user[0] == '\0' || group[0] == '\0')
    {
        LOG(LOG_LEVEL_ERROR,
            "Both a runtime_user and a runtime_group MUST be specified");
    }
    else
    {
        rv = g_drop_privileges(user, group);
        if (rv == 0)
        {
            LOG(LOG_LEVEL_INFO, "Switched user:group to %s:%s", user, group);
        }
    }

    return rv;
}
```

On a Linux system the process should probably check something like the ownership of  `/proc/self` to verify if it's really running as root before either trying to drop privileges or printing scary warnings about running as root. I think this should fix the issue without having to think more carefully about the init system/daemon management.

--- a/xrdp/xrdp.c
+++ b/xrdp/xrdp.c
@@ -446,7 +446,7 @@ check_drop_privileges(struct xrdp_startup_params *startup_params)
         // We're running as a privileged user, and this isn't
         // overridden in our own config
         LOG(LOG_LEVEL_ALWAYS,
-            "You are running xrdp as uid=%d gid=%d. This is not safe.",
+            "You are running xrdp as uid=%d gid=%d. This is not recommended.",
             entry_uid, entry_gid);
         rv = 0; // Allowed for now for old configs
     }

--- a/xrdp/xrdp.c
+++ b/xrdp/xrdp.c
@@ -426,13 +426,29 @@ check_drop_privileges(struct xrdp_startup_params *startup_params)
     int rv = 1;
     const char *user = startup_params->runtime_user;
     const char *group = startup_params->runtime_group;
+    int entry_uid = g_getuid();
+    int entry_gid = g_getgid();
 
-    if (user[0] == '\0' && group[0] == '\0')
+    // Regard a user as privileged if either the UID or GID is zero
+    int privileged_user = (entry_uid == 0 || entry_gid == 0);
+    if (!privileged_user)
     {
-        // Allow this for now
-        LOG(LOG_LEVEL_ALWAYS,
-            "You are running xrdp as root. This is not safe.");
+        // xrdp daemon UID:GID is already set by whatever started it. We'll
+        // log this as it may be unexpected and continue. We don't have
+        // any privilege to override this.
+        LOG(LOG_LEVEL_INFO,
+            "xrdp is already running as an unprivileged user uid=%d gid=%d",
+            entry_uid, entry_gid);
         rv = 0;
+    }
+    else if (user[0] == '\0' && group[0] == '\0')
+    {
+        // We're running as a privileged user, and this isn't
+        // overridden in our own config
+        LOG(LOG_LEVEL_ALWAYS,
+            "You are running xrdp as uid=%d gid=%d. This is not safe.",
+            entry_uid, entry_gid);
+        rv = 0; // Allowed for now for old configs
     }
     else if (user[0] == '\0' || group[0] == '\0')
     {

